src/essreflectometry/orso.py

```diff
@@ -1,15 +1,165 @@
+import os
+import platform
+from datetime import datetime, timezone
+from typing import NewType, Optional
 
+import sciline
+from dateutil.parser import parse as parse_datetime
+from orsopy.fileio import base as orso_base
+from orsopy.fileio import data_source, reduction
 
-def not_found_warning():
+from .types import Filename, RawData, Reference, Run, Sample
+
+
+class OrsoExperiment(
+    sciline.Scope[Run, data_source.Experiment], data_source.Experiment
+):
+    """ORSO experiment for a run."""
+
+
+class OrsoInstrument(
+    sciline.Scope[Run, data_source.InstrumentSettings], data_source.InstrumentSettings
+):
+    """ORSO instrument settings for a run."""
+
+
+class OrsoOwner(sciline.Scope[Run, orso_base.Person], orso_base.Person):
+    """ORSO owner of a file."""
+
+
+class OrsoReduction(sciline.Scope[Run, reduction.Reduction], reduction.Reduction):
+    """ORSO measurement for a run."""
+
+
+class OrsoSample(sciline.Scope[Run, data_source.Sample], data_source.Sample):
+    """ORSO sample of a run."""
+
+
+OrsoCreator = NewType('OrsoCreator', orso_base.Person)
+"""ORSO creator, that is, the person who processed the data."""
+
+OrsoDataSource = NewType('OrsoDataSource', data_source.DataSource)
+"""ORSO data source."""
+
+OrsoMeasurement = NewType('OrsoMeasurement', data_source.Measurement)
+"""ORSO measurement."""
+
+
+def parse_orso_experiment(raw_data: RawData[Run]) -> OrsoExperiment[Run]:
+    """Parse ORSO experiment metadata from raw NeXus data."""
+    return OrsoExperiment(
+        data_source.Experiment(
+            title=raw_data['title'],
+            instrument=raw_data['name'],
+            facility=raw_data['facility'],
+            start_date=parse_datetime(raw_data['start_time']),
+            probe='neutron',
+        )
+    )
+
+
+def parse_orso_owner(raw_data: RawData[Run]) -> OrsoOwner[Run]:
+    """Parse ORSO owner metadata from raw NeXus data."""
+    return OrsoOwner(
+        orso_base.Person(
+            name=raw_data['user']['name'],
+            contact=raw_data['user']['email'],
+            affiliation=None,
+        )
+    )
+
+
+def parse_orso_sample(raw_data: RawData[Run]) -> OrsoSample[Run]:
+    """Parse ORSO sample metadata from raw NeXus data."""
+    if not raw_data.get('sample'):
+        return OrsoSample(data_source.Sample.empty())
+    raise NotImplementedError('NeXus sample parsing is not implemented')
+
+
+def build_orso_measurement(
+    sample_filename: Filename[Sample],
+    reference_filename: Filename[Reference],
+    instrument: Optional[OrsoInstrument[Sample]],
+) -> OrsoMeasurement:
+    """Assemble ORSO measurement metadata."""
+    # TODO populate timestamp
+    #      doesn't work with a local file because we need the timestamp of the original,
+    #      SciCat can provide that
+    return OrsoMeasurement(
+        data_source.Measurement(
+            instrument_settings=instrument,
+            data_files=[orso_base.File(file=os.path.basename(sample_filename))],
```

Comment:
This seems to be a limitation of the Orso standard, so probably little we can do about it. But, identifying a file by the file name seems an unreliable way to achieve reproducibility.
Best way would of course be to store the file itself, second best would be to store a hash and preferably also instructions on how to access the file.

Do you think there is anything we can do about that? Or do we just follow the standard for now? I looked quickly at the orso standard spec but could not find any `metadata` field where that kind of extra information could be stored.

Reply:
ORSO is not like NeXus where you can just write whatever you want. So we are limited by the standard. The only thing we can do (in the short term) is to store any additional info in the comment field. As I understand, this is a free form text filed. But it's meant to be read by humans, not programs...

If you think this is important enough (and I think it is), consider raising it with ORSO: https://github.com/reflectivity/file_format

Reply:
As a temporary solution we could just append ` sha256:fff4b52fd585a4ca3e9709a64da93209f32db841a07ff7a1cde12db2f8bfe0a3` (or maybe md5 but it probably doesn't matter) to the end of the comment. I'd happily raise this with ORSO.

Reply:
https://github.com/reflectivity/file_format/issues/15